src/EventIndexing.js

```diff
@@ -0,0 +1,406 @@
+                    this.crawlerCheckpoints.push(backCheckpoint);
+                    this.crawlerCheckpoints.push(forwardCheckpoint);
+                }));
+            };
+
+            // If our indexer is empty we're most likely running Riot the
+            // first time with indexing support or running it with an
+            // initial sync. Add checkpoints to crawl our encrypted rooms.
+            const eventIndexWasEmpty = await indexManager.isEventIndexEmpty();
+            if (eventIndexWasEmpty) await addInitialCheckpoints();
+
+            // Start our crawler.
+            this.startCrawler();
+            return;
+        }
+
+        if (prevState === "SYNCING" && state === "SYNCING") {
+            // A sync was done, presumably we queued up some live events,
+            // commit them now.
+            console.log("EventIndex: Committing events");
+            await indexManager.commitLiveEvents();
+            return;
+        }
+    }
+
+    async onRoomTimeline(ev, room, toStartOfTimeline, removed, data) {
+        const indexManager = PlatformPeg.get().getEventIndexingManager();
+        if (indexManager === null) return;
+
+        // We only index encrypted rooms locally.
+        if (!MatrixClientPeg.get().isRoomEncrypted(room.roomId)) return;
+
+        // If it isn't a live event or if it's redacted there's nothing to
+        // do.
+        if (toStartOfTimeline || !data || !data.liveEvent
+            || ev.isRedacted()) {
+            return;
+        }
+
+        // If the event is not yet decrypted mark it for the
+        // Event.decrypted callback.
+        if (ev.isBeingDecrypted()) {
+            const eventId = ev.getId();
+            this.liveEventsForIndex.add(eventId);
+        } else {
+            // If the event is decrypted or is unencrypted add it to the
+            // index now.
+            await this.addLiveEventToIndex(ev);
+        }
+    }
+
+    async onEventDecrypted(ev, err) {
+        const indexManager = PlatformPeg.get().getEventIndexingManager();
+        if (indexManager === null) return;
+
+        const eventId = ev.getId();
+
+        // If the event isn't in our live event set, ignore it.
+        if (!this.liveEventsForIndex.delete(eventId)) return;
+        if (err) return;
+        await this.addLiveEventToIndex(ev);
+    }
+
+    async addLiveEventToIndex(ev) {
+        const indexManager = PlatformPeg.get().getEventIndexingManager();
+        if (indexManager === null) return;
+
+        if (["m.room.message", "m.room.name", "m.room.topic"]
+            .indexOf(ev.getType()) == -1) {
+            return;
+        }
+
+        const e = ev.toJSON().decrypted;
+        const profile = {
+            displayname: ev.sender.rawDisplayName,
+            avatar_url: ev.sender.getMxcAvatarUrl(),
+        };
+
+        indexManager.addEventToIndex(e, profile);
+    }
+
+    async crawlerFunc(handle) {
```

Comment:
Maybe drop the `Func`?

```suggestion
    async crawl(handle) {
```